File: libs/sdk-dotnet/AuthzSdk/Exceptions/AuthzClientException.cs
namespace AuthzSdk.Exceptions;

public sealed class AuthzClientException : Exception
{
    public AuthzClientException(string message)
        : this(message, statusCode: 0, innerException: null)
    {
    }

    public AuthzClientException(string message, Exception innerException)
        : this(message, statusCode: 0, innerException)
    {
    }

    public AuthzClientException(string message, int statusCode)
        : this(message, statusCode, innerException: null)
    {
    }

    public AuthzClientException(string message, int statusCode, Exception? innerException)
        : base(message, innerException)
    {
        StatusCode = statusCode;
    }

    public int StatusCode { get; }

    public bool IsClientError => StatusCode is >= 400 and < 500;

    public bool IsServerError => StatusCode >= 500;
}

File: libs/sdk-java/src/main/java/com/platform/authz/sdk/exception/AuthzClientException.java
package com.platform.authz.sdk.exception;

/**
 * Exception thrown by the AuthZ SDK when a remote call fails.
 * Wraps HTTP errors, timeouts, and circuit-breaker open states.
 */
public class AuthzClientException extends RuntimeException {

    private final int statusCode;

    public AuthzClientException(String message) {
        super(message);
        this.statusCode = 0;
    }

    public AuthzClientException(String message, Throwable cause) {
        super(message, cause);
        this.statusCode = 0;
    }

    public AuthzClientException(String message, int statusCode) {
        super(message);
        this.statusCode = statusCode;
    }

    public AuthzClientException(String message, int statusCode, Throwable cause) {
        super(message, cause);
        this.statusCode = statusCode;
    }

    public int getStatusCode() {
        return statusCode;
    }

    public boolean isClientError() {
        return statusCode >= 400 && statusCode < 500;
    }

    public boolean isServerError() {
        return statusCode >= 500;
    }
}
